modules/seqexec/engine/src/main/scala/seqexec/engine/Engine.scala

```diff
@@ -344,17 +330,8 @@ class Engine[D, U](stateL: Lens[D, Engine.State]) {
 
 object Engine {
 
-  final case class State(sequences: Map[Observation.Id, Sequence.State[IO]])
-
-  object State {
-    def empty: State = State(Map.empty)
-
-    def sequences: Lens[State, Map[Observation.Id, Sequence.State[IO]]] =
-      Lens[State, Map[Observation.Id, Sequence.State[IO]]](_.sequences)(ss => _ => State.apply(ss))
-    private def atSequence(id: Observation.Id)
-    : Optional[Map[Observation.Id, Sequence.State[IO]], Sequence.State[IO]] = index(id)
-    def sequenceState(id: Observation.Id): Optional[State, Sequence.State[IO]] =
-      sequences ^|-? atSequence(id)
+  trait State[D] {
+    def sequenceStateIndex(sid: Observation.Id): Optional[D, Sequence.State[IO]]
```

Comment:
This is the core change. Instead of a `case class` containing the engine view of the sequences, `State` is now a trait that must implement a Lens to the sequences. So now the container can keep the engine sequence with its own sequence data, which is invisible to the engine. The only complication I found was with the creation and destruction of the sequences at the engine level.